src/main/kotlin/com/coder/gateway/sdk/CoderRestClientService.kt

```diff
@@ -218,3 +250,203 @@ class CoderRestClient(var url: URL, var token: String,
+    certChain.withIndex().forEach {
+        truststore.setCertificateEntry("cert${it.index}", it.value as X509Certificate)
+    }
+    trustManagerFactory.init(truststore)
+    return trustManagerFactory.trustManagers.map { MergedSystemTrustManger(it as X509TrustManager) }.toTypedArray()
+}
+
+fun expandPath(path: String): String {
+    if (path.startsWith("~/")) {
+        return Path.of(System.getProperty("user.home"), path.substring(1)).toString()
+    }
+    if (path.startsWith("\$HOME/")) {
+        return Path.of(System.getProperty("user.home"), path.substring(5)).toString()
+    }
+    if (path.startsWith("\${user.home}/")) {
+        return Path.of(System.getProperty("user.home"), path.substring(12)).toString()
+    }
+    return path
+}
+
+class AlternateNameSSLSocketFactory(private val delegate: SSLSocketFactory, private val alternateName: String) : SSLSocketFactory() {
+    override fun getDefaultCipherSuites(): Array<String> {
+        return delegate.defaultCipherSuites
+    }
+
+    override fun getSupportedCipherSuites(): Array<String> {
+        return delegate.supportedCipherSuites
+    }
+
+    override fun createSocket(): Socket {
+        val socket = delegate.createSocket() as SSLSocket
+        customizeSocket(socket)
+        return socket
+    }
+
+    override fun createSocket(host: String?, port: Int): Socket {
+        val socket = delegate.createSocket(host,  port) as SSLSocket
+        customizeSocket(socket)
+        return socket
+    }
+
+    override fun createSocket(host: String?, port: Int, localHost: InetAddress?, localPort: Int): Socket {
+        val socket = delegate.createSocket(host, port, localHost, localPort) as SSLSocket
+        customizeSocket(socket)
+        return socket
+    }
+
+    override fun createSocket(host: InetAddress?, port: Int): Socket {
+        val socket = delegate.createSocket(host, port) as SSLSocket
+        customizeSocket(socket)
+        return socket
+    }
+
+    override fun createSocket(address: InetAddress?, port: Int, localAddress: InetAddress?, localPort: Int): Socket {
+        val socket = delegate.createSocket(address, port, localAddress, localPort) as SSLSocket
+        customizeSocket(socket)
+        return socket
+    }
+
+    override fun createSocket(s: Socket?, host: String?, port: Int, autoClose: Boolean): Socket {
+        val socket = delegate.createSocket(s, host, port, autoClose) as SSLSocket
+        customizeSocket(socket)
+        return socket
+    }
+
+    private fun customizeSocket(socket: SSLSocket) {
+        val params = socket.sslParameters
+        params.serverNames = listOf(SNIHostName(alternateName))
+        socket.sslParameters = params
+    }
+}
+
+class CoderHostnameVerifier(private val alternateName: String) : HostnameVerifier {
+    override fun verify(host: String, session: SSLSession): Boolean {
+        if (alternateName.isEmpty()) {
+            println("using default hostname verifier, alternateName is  empty")
```

Comment:
We should copy the logger pattern used elsewhere instead of using `println` directly.